Ruins/src/main/java/atomicstryker/ruins/common/RuinVennCriterion.java

```diff
@@ -0,0 +1,656 @@
+package atomicstryker.ruins.common;
+
+import java.util.ArrayDeque;
+import java.util.Collection;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import javax.annotation.Nonnull;
+
+import org.apache.commons.lang3.StringEscapeUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Criterion against which collections of strings can be tested. The conditions to be met in order to satisfy a
+ * particular criterion are defined by an expression provided at the time of its creation.
+ * <p>
+ * An expression consists of a number of <i>elements</i>, or strings, to be found (or not) among collections compared
+ * to it. Elements are case sensitive and must either consist entirely of alphanumeric characters and/or underscores or
+ * be enclosed in quotes. Quoted elements may contain standard Java escape sequences. Elements are combined to form
+ * expressions through the use of the following <i>operators</i>:
+ * <ul>
+ * <li>unary minus, representing complement ("-A" means any collection <i>not</i> containing A)</li>
+ * <li>binary plus, representing intersection ("A+B" means any collection containing both A <i>and</i> B)</li>
+ * <li>binary minus, representing relative complement ("A-B" means any collection containing A <i>and not</i> B)</li>
+ * <li>binary comma, representing union ("A,B" means any collection containing either A <i>or</i> B, or both)</li>
+ * </ul>
+ * Unary plus ("+A") is also accepted as an operator for completeness, but it has no effect. The unary operators have
+ * highest precedence, followed by the binary plus and minus operators, and finally the comma. Order of evaluation can
+ * also be modified by using parentheses to enclose subexpressions. Whitespace is ignored, as is the number sign and
+ * anything following it (unless enclosed in quotes).
+ * <p>
+ * Example: A criterion created from the expression <code>RED + (YELLOW, GREEN) - BLUE</code> is satisfied by any
+ * collection of strings containing RED, either TELLOW or GREEN (or both), but not BLUE.
+ * <p>
+ * Optionally, a collection of elements may also be provided from which an expression must be constructed. If an
+ * element in the expression does not appear in the list of valid elements, an exception is thrown (if strict
+ * validation is enabled) or a warning is logged (if strict validation is not enabled).
+ */
+class RuinVennCriterion
```

Comment:
Why "Ruin", for starters it would be "Ruins" but also there is nothing in here specific to Ruins

Reply:
Half the files start with "Ruin"...I thought that was your naming convention. "Ruins" only appears in one file name: RuinsMod. I'm happy to change it if you'd like.

Reply:
There's a difference between a Ruin and Ruins, the mod :)

Reply:
Not to say i'm not inconsistent, because i probably am